[WIP] proposed new unicode module

jasnell opened the pull request:
The Node.js enhancement proposal https://github.com/nodejs/node-eps/pull/15 proposes the addition of a new `'unicode'` module that introduces a number of ICU backed APIs, including the ability to transcode a Buffer from one encoding to another. This is the prototype implementation of the new module.

It is a **work in progress**. This is not intended to be landed at this time.

/cc @trevnorris @addaleax @srl295

mscdex commented on doc/api/unicode.md:
```diff
@@ -0,0 +1,55 @@
+Returns the character located at the given offset in the `Buffer`. Works even
+if the offset falls in the middle of a multibyte UTF-8 or UTF-16 character.
+
+## unicode.utf8Slice(buf, start, end)
```
Why restrict this to just utf8? What about just having general `slice()` and `length()` functions that accept an encoding name. I'm thinking this would be beneficial for other multi-byte encodings as well, such as utf16.

jasnell replied:
It's just UTF8 for now, I'm going to expand this out to support each of our main encodings.

jasnell replied:
that said, out of the encodings we support out of the box (utf8, utf16le, latin1, and ascii), utf8 is the only one that's difficult here. latin1 and ascii would fallback to Buffer.prototype.split and Buffer.prototype.length, and utf16le is always 2 bytes so length is always Buffer.prototype.length / 2 and split is straightforward.

Where it begins to get complicated is if we decide to support the full range of encodings offered by ICU with full-data.

yorkie commented on doc/api/unicode.md:
````diff
@@ -0,0 +1,55 @@
+const unicode = require('unicode');
+```
+
+## unicode.transcode(buf, from_enc, to_enc)
````
`fromEncoding` and `toEncoding` might be like a argument in this community :-)

yorkie commented on lib/unicode.js:
```diff
@@ -0,0 +1,210 @@
+  };
+
+  exports.transcode = function transcode(source, from_enc, to_enc, options) {
+    if (!source) return;
```
Why the source is not defined, it just returns undefined, but `source.buffer` is not an `ArrayBuffer`, the program throws a `TypeError`, use `throw` instead of `return` might be reasonable here?

yorkie commented on doc/api/unicode.md:
```diff
@@ -0,0 +1,55 @@
+* `to_enc` {string} The target encoding
+
+Re-encodes the given `Buffer` from one character encoding to another. Returns
+a new `Buffer` instance.
```
This description here is not accurate, [when `from_enc` and `to_enc` is same](https://github.com/jasnell/node/blob/842e0810de3ebdba739695c259dc9b8c8a58f840/lib/unicode.js#L70-L71), this function just returns the original `source` instance, not a new instance of `Buffer`.

_Suggestion: remove the "new" adjective._

targos replied:
It may be less surprising if it was always returning a new Buffer. Like `Buffer.concat` when the list has only one Buffer.

yorkie commented on lib/unicode.js:
```diff
@@ -0,0 +1,210 @@
+      if (cnv_to)
+        return cnv_to(source, options.lenient);
+    }
+    throw new Error(`Unsupported conversion: ${from_enc} to ${to_enc}`);
```
Is `TypeError` here reasonable? :-)

yorkie commented on doc/api/unicode.md:
```diff
@@ -0,0 +1,55 @@
+* `buf` {Buffer} A `Buffer` instance
+* `from_enc` {string} The current encoding
+* `to_enc` {string} The target encoding
+
```
`options` argument is missing :-)

yorkie commented on lib/unicode.js:
```diff
@@ -0,0 +1,210 @@
+    if (from_enc === to_enc)
+      return source;
+
+    options = options || kDefaultOptions;
```
Should we use `util._extend` to avoid changing unexpectedly the default `kDefaultOptions` in below calls?

yorkie commented on lib/unicode.js:
```diff
@@ -0,0 +1,210 @@
+      case 'utf16-le':
+        return 'ucs2';
+      default:
+        return enc;
```
Propose to throw instead of returning, throwing might be more easy to notify the wrong inputs from user-land.

```js
default: 
  throw new TypeError(`encoding ${enc} is invalid`);
```

yorkie commented on src/node_i18n.cc:
```diff
@@ -185,13 +219,379 @@ static void ToASCII(const FunctionCallbackInfo<Value>& args) {
+  } else {
+    if (IsBigEndian()) {
+      const uint32_t len = ts_obj_length >> 1;
+      UChar* copy = static_cast<UChar*>(malloc(len));
```
I guess we should check if `malloc` is correct running :-)

addaleax replied:
Ack… alternatively, maybe use `MaybeStackBuffer<UChar> copy(len);` here, too? (and basically ditto for all following `malloc()` calls). Also, this seems too short? Shouldn’t it be `malloc(len << 1)`?

addaleax commented on doc/api/unicode.md:
```diff
@@ -0,0 +1,55 @@
+# icu
```
`unicode`? Also, it might be good to add this to `exports.builtinLibs` in `lib/internal/module.js`.

addaleax commented on doc/api/unicode.md:
```diff
@@ -0,0 +1,55 @@
+# icu
+
+    Stability: 1 - Experimental
```
Since c809b8834541bf63b1a43d5a12833d98ffba834f we’re using blockquotes here (`> Stability: 1 - Experimental`) instead of indentation

addaleax commented on doc/api/unicode.md:
```diff
@@ -0,0 +1,55 @@
+Re-encodes the given `Buffer` from one character encoding to another. Returns
+a new `Buffer` instance.
+
+## unicode.codePointAt(buf, pos, encoding)
```
I will mistype this as `charCodeAt` so often… ;)

addaleax commented on doc/api/unicode.md:
```diff
@@ -0,0 +1,55 @@
+## unicode.codePointAt(buf, pos, encoding)
+
+* `buf` {Buffer} A `Buffer` instance
+* `pos` {integer} The offset position
```
I would add `in bytes` or something similar here to make clear that this is not measured in characters.

addaleax commented on lib/unicode.js:
```diff
@@ -0,0 +1,210 @@
+'use strict';
+
+const has_icu = process.binding('config').hasIntl;
```
Maybe this boolean should be on `exports`?

targos replied:
nit: camelCase

addaleax commented on src/node_i18n.cc:
```diff
@@ -51,17 +59,43 @@ extern "C" const char U_DATA_API SMALL_ICUDATA_ENTRY_POINT[];
 
 namespace i18n {
 
+#define THROW_AND_RETURN_UNLESS_BUFFER(env, obj)                              \
```
How about moving these macros into `util.h`? I think I would prefer that over copypasting.

addaleax commented on src/node_i18n.cc:
```diff
@@ -185,13 +219,379 @@ static void ToASCII(const FunctionCallbackInfo<Value>& args) {
                           len).ToLocalChecked());
 }
 
+// Get's the codepoint at a given offset for UTF-8 or UCS2
```
typo: `Gets` … ;)

addaleax commented on src/node_i18n.cc:
```diff
@@ -185,13 +219,379 @@ static void ToASCII(const FunctionCallbackInfo<Value>& args) {
+  size_t len;
+
+  MaybeStackBuffer<UChar> buf;
+  buf.AllocateSufficientStorage(ts_obj_length);
```
You can pass `ts_obj_length` to the constructor, if you prefer.

addaleax commented on src/node_i18n.cc:
```diff
@@ -185,13 +219,379 @@ static void ToASCII(const FunctionCallbackInfo<Value>& args) {
+    source = reinterpret_cast<UChar*>(ts_obj_data);
+  } else {
+    swapspace.AllocateSufficientStorage(length);
+    source = static_cast<UChar*>(malloc(ts_obj_length >> 1));
```
`source` is overridden below with `*swapspace`, I think this line does nothing but leak memory.

addaleax commented:
Nice, I’m excited! I’ve PRed with some of the changes I’m suggesting here at jasnell/node#10, feel free to do with those whatever you like!

jasnell commented:
Awesome, thanks all. The impl here is still pretty rough and I already have some updates queued up. I'll work in the suggested changes and I'll take a look at your pr @addaleax. Much appreciated!

targos commented on doc/api/unicode.md:
```diff
@@ -0,0 +1,55 @@
+* `encoding` The character encoding of the `Buffer` data. Default = `'utf8'`
+
+Returns the Unicode codepoint located at the given offset in the `Buffer`. Works
+even if the offset falls in the middle of a multibyte UTF-8 or UTF-16 character.
```
does it mean it returns the same codepoint if you are at the start or in the middle of a UTF-8 character?
If that is the case, it may be good to say that the behavior is different than String.codePointAt

targos commented on doc/api/unicode.md:
```diff
@@ -0,0 +1,55 @@
+* `encoding` The character encoding of the `Buffer` data. Default = `'utf8'`
+
+Returns the character located at the given offset in the `Buffer`. Works even
+if the offset falls in the middle of a multibyte UTF-8 or UTF-16 character.
```
ditto (emphasize the difference with String.charAt)

targos commented:
The name of this module conflicts with https://www.npmjs.com/package/unicode. How do you plan to handle this ?

jasnell commented:
@targos ... yeah, because of the name conflict I renamed this back to `icu` instead of `unicode`.

@addaleax ... updated! I had already started making a number of similar changes so I just manually worked the rest of your suggestions in and squashed everything down. Also added a test case and tweaked the documentation a bit. PTAL!

addaleax commented on doc/api/icu.md:
```diff
@@ -0,0 +1,65 @@
+## icu.codePointAt(buf, offset, encoding)
+
+* `buf` {Buffer} A `Buffer` instance
+* `offset` {integer} The offset position
```
I still think these should have clarified that `offset` is measured in bytes, not characters, even if it’s just saying `byte offset`.

jasnell replied:
oh! right, missed that one. Will add that

jasnell commented:
Also, something to keep in mind, ICU exposes a significant amount of additional functionality. This PR just scratches the surface. Other things we can look at for this include:

1. UTF8 and UTF16 codepoint/character iteration for Buffer.. for instance (`for (cp of icu.codepoints(buf, 'utf8')) { ... }` or `for (ch of icu.chars(buf, 'utf8') { ... }`)... which would allow efficient iteration over a Buffer instance without first having to do a toString() on it.

2. Unicode Character Properties.. that is, retrieve the character properties for any given Unicode codepoint.

3. Locale details.. information about Locales, basically. This would provide additional detail not currently covered by TC39-402 proposals to avoid overlap.

4. Searching Buffers for substrings.. UTF8 and UTF16 aware searching of Buffers without first having to do a toString() on those.

trevnorris commented:
@jasnell would this be exposed to the native API?

trevnorris commented on doc/api/icu.md:
```diff
@@ -0,0 +1,76 @@
+* `to_enc` {string} The target encoding
+
+Re-encodes the given `Buffer` from one character encoding to another. Returns
+a new `Buffer` instance.
```
would it be possible to allow an optional `buffer` argument so another one doesn't need to be created? not for this PR, but would be a nice addition in the future.

jasnell replied:
Potentially. The key issue with this is that the returned Buffer could potentially be quite a bit larger than the source so we would have to decide the acceptable behavior if the passed in Buffer is not large enough. Would we throw or truncate?

trevnorris replied:
dunno. probably throw.

jasnell replied:
I will explore the possibility of doing this for all of the various APIs that return a Buffer in a future iteration.

trevnorris commented on doc/api/icu.md:
```diff
@@ -0,0 +1,76 @@
+
+* `buf` {Buffer} A `Buffer` instance
+* `offset` {integer} The byte offset position
+* `encoding` The character encoding of the `Buffer` data. Default = `'utf8'`
```
Having a "Default" implies it's optional. So the signature should be
```
icu.codePointAt(buf, offset[, encoding])
```

trevnorris commented on doc/api/icu.md:
```diff
@@ -0,0 +1,76 @@
+
+* `buf` {Buffer} A `Buffer` instance
+* `offset` {integer} The byte offset position
+* `encoding` The character encoding of the `Buffer` data. Default = `'utf8'`
```
Same thing with "Default", and preemptively mentioning for any other locations.

trevnorris commented on doc/api/icu.md:
```diff
@@ -0,0 +1,76 @@
+arguments define character offsets rather than byte offsets. Ensures that the
+slice occurs at proper codepoint boundaries.
+
+## icu.utf8Slice(buf, start, end)
```
what's the difference between this and simply passing `'utf8'` to `.slice()`?

jasnell replied:
`icu.slice(buf, 'utf8')` defers to `icu.utf8Slice()`... so there would be no difference. I'll likely be pulling `icu.utf8Slice()` off the exported API.

trevnorris commented on doc/api/icu.md:
````diff
@@ -0,0 +1,76 @@
+functionality provided by the ICU4C library.
+
+```js
+const icu = require('icu');
````
I remember you mentioning there was a function to tell you if the buffer is one byte or not, or am I misremembering something?

jasnell replied:
ICU includes charset detection APIs that I had in here originally but opted to pull back out based on the feedback.

trevnorris commented on lib/icu.js:
```diff
@@ -0,0 +1,199 @@
+
+// Maps the supported transcoding conversions. The top key is the from_enc,
+// the child key is the to_enc. The value is the transcoding function to.
+const conversions = {
```
This cover all the encodings that `Buffer.isEncoding()` returns `true` for?

jasnell replied:
It does not handle `hex` or `base64`. It should handle all the others. `internalUtil.normalizeEncoding()` is used under the covers.

trevnorris replied:
sorry. what I meant is does it handle the variations of supported APIs (e.g. should still probably support `'binary'` for the time being; and also both `utf8` and `utf-8`).

jasnell replied:
Yep, but I'll be expanding the test case to make absolutely certain.

trevnorris replied:
sorry. not following what you'll be absolutely certain of.

jasnell replied:
that all of the various aliases of our core encodings are supported

trevnorris commented on lib/icu.js:
```diff
@@ -0,0 +1,199 @@
+// based on the given encoding
+exports.codePointAt = function(buf, pos, encoding) {
+  if (!buf || !(buf.buffer instanceof ArrayBuffer))
+    throw new TypeError('First argument must be a Buffer');
```
this necessary as it's the first check in C++, and the fact that this can be fooled fairly easily?

jasnell replied:
Yeah, I was thinking of pulling this check back out.

jasnell commented:
@trevnorris ... re: the native API, I was going to be working on making sure that the ICU headers were included and that native modules could be linked against the ICU bundled in Node. If that works, then there would be little reason to expose this module directly to native addons since they could go directly to the ICU API.

trevnorris commented on src/node_i18n.cc:
```diff
@@ -185,13 +211,338 @@ static void ToASCII(const FunctionCallbackInfo<Value>& args) {
+
+  if (U_SUCCESS(status)) {
+    args.GetReturnValue().Set(
+      Buffer::Copy(env->isolate(), *buf, len).ToLocalChecked());
```
there a way to grab the pointer on `buf` if the allocation was larger than `kStackStorageSize` instead of doing a `memcpy()` every time?

jasnell replied:
We could, yes...

trevnorris commented:
Like where this is going. Thanks for doing this.

jasnell commented:
Updated to address nits/feedback. Also, added a `icu.normalize(buf, form)` API to allow for unicode normalization of Buffers without having to toString first.

jasnell commented:
@trevnorris ... in the next round of updates I will work on the bit about avoiding Buffer::copy where possible

jasnell commented:
Ok, added a change that allows a Buffer to be created directly from a MaybeStackBuffer. If the stack buffer is not-allocated, then the Buffer copies the data; otherwise the data is used without copying and the MaybeStackBuffer is released from responsibility for freeing the data (the Buffer instance takes over). May not be perfect so I'd appreciate a look @trevnorris and @addaleax!

addaleax commented on src/node_buffer.h:
```diff
@@ -47,6 +47,11 @@ NODE_EXTERN v8::MaybeLocal<v8::Object> New(v8::Isolate* isolate,
+// public constructor - data is used if memory was allocated, copied otherwise.
+NODE_EXTERN v8::MaybeLocal<v8::Object> New(v8::Isolate* isolate,
+                                           MaybeStackBuffer<char>* buf,
+                                           size_t len);
```
I think this would be easier to integrate in `MaybeStackBuffer` as a method. Imo it kind of makes sense there, and on the more technical side, `node_buffer.h` is a public header while `util.h` is an internal one.

jasnell replied:
Ok.. perhaps something like a `msb.AsBuffer(env->isolate())` on `MaybeStackBuffer`?

addaleax replied:
Yep, sounds good.

addaleax commented on src/util.h:
```diff
@@ -341,15 +341,30 @@ class MaybeStackBuffer {
   size_t length_;
   T* buf_;
   T buf_st_[kStackStorageSize];
+  bool released_;
```
You could probably just assign `buf_` to `buf_st_` when transferring the memory ownership. If you want to stick with a property, I’d suggest moving it in front of `buf_st_` so that `length_`, `buf_`, `released_` (and in the common case, the actual data, too) fit into a single cache line.

jasnell replied:
Yep.. would need to make sure Release() is called after (instead of before) like I have it now, but this is goodness.

trevnorris commented on lib/icu.js:
```diff
@@ -0,0 +1,209 @@
+  encoding = normalizeEncoding(encoding || 'utf8');
+
+  if (encoding === 'binary' || encoding === 'ascii')
+    return String.fromCharCode(buf[pos]);
```
Should we do any coercion if encoding is 'ascii' but the code point lies outside the correct range?

trevnorris replied:
Also, shouldn't it be `'latin1'` and not `'binary'` based on all the other code?

jasnell replied:
Good catch! I originally wrote this before that change was made ;-)

jasnell replied:
Regarding what to do with non-ascii bytes... replacing with `?` would likely be the best choice.

trevnorris commented on lib/icu.js:
```diff
@@ -0,0 +1,209 @@
+    throw new TypeError('First argument must be a Buffer');
+  if (typeof end === 'undefined') end = buf.length;
+  start >>= 0;
+  end >>= 0;
```
This doesn't operate like `String#slice()` or `Buffer#slice()`. If a negative number is passed then it should slice going backwards. There are some other minor argument coercion differences between this and how Typed Array handles arguments. Buffer should also have its argument parsing reviewed and fixed, but might as well start now. Here's an idea of how they should be parsed:
```js
if (!start) start = 0;
// This works b/c "start < 0" calls valueOf() from start. Evaluate this
// way to be consistent with how Typed Array evaluates. For example:
//   ui8array.subarray(4.6, 5.1): 4.6 -> 4; 5.1 -> 5
//   ui8array.subarray(-3.1, -1.9): -3.1 -> -3; -1.9 -> -1
else start = start < 0 ? Math.ceil(start) : Math.floor(start);

if (end === undefined) end = buf.length;
else if (!end) end = 0;
else end = end < 0 ? Math.ceil(end) : Math.floor(end);

if (end <= start) return Buffer.alloc(0);
```

There's probably be a simpler way to achieve this, but intentionally wanted this to be verbose to easily see what's going on.

jasnell replied:
Yeah, I've had a standing todo to go back and reconcile this. I'll hit that in the next update.

jasnell commented:
My plan on this is to back off adding a new module. And instead:

a) Have separate PRs that add transcode and normalize as statics on require('buffer') ... NOT as prototype methods on `Buffer` but as statics on the module itself; and 
b) To introduce getCharacterProperty() and getColumnWidth() in separate PRs.

jasnell commented:
Closed this so I can work on breaking it up into individual PRs
